File: 23. Merge k Sorted Lists/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> pq = new PriorityQueue<>((a, b) -> a.val - b.val);
        for (ListNode curr : lists) {
            while (curr != null) {
                ListNode foll = curr.next;
                pq.offer(curr);
                curr = foll;
            }
        }
        if (pq.isEmpty()) return null;
        ListNode root = pq.poll(), curr = root;
        while (!pq.isEmpty()) {
            curr.next = pq.poll();
            curr = curr.next;
        }
        curr.next = null;
        return root;
    }
}

File: 23. Merge k Sorted Lists/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        pq = []
        i = 0
        for curr in lists:
            while curr != None:
                heapq.heappush(pq, (curr.val, i, curr))
                curr = curr.next
                i += 1
        if not pq: return None
        root = heapq.heappop(pq)[2]
        curr = root
        while pq:
            curr.next = heapq.heappop(pq)[2]
            curr = curr.next
        curr.next = None
        return root